src/apps/appstore/app.js

```diff
@@ -393,8 +513,274 @@ define(function(require) {
 					$('#appstore_container .app-filter.active').click();
 				});
 			});
-		}
+		},
+
+		showMarketplaceConnector: function(parent) {
+			const self = this;
+			const appstoreData = self.getStore('appstoreData');
+			const marketConfig = appstoreData.marketplace || {};
+			const enabled = marketConfig.enabled;
+			const is_linked = marketConfig.is_linked;
+
+			if (enabled && is_linked) {
+				self.renderMarketSettings(parent);
+			} else if (enabled) {
+				self.renderMarketLink(parent);
+			} else {
+				self.renderMarketWelcome(parent);
+			}
+		},
+
+		renderMarketLink: function(parent) {
+			const self = this;
+			const appstoreData = self.getStore('appstoreData');
+			const marketConfig = appstoreData.marketplace;
+			const template = $(self.getTemplate({
+				name: 'marketClusterLink',
+				data: {
+					config: marketConfig
+				}
+			}));
+
+			monster.ui.insertTemplate(parent.find('.right-container'), template);
+			self.bindMarketLinkEvents(parent, template);
+		},
+
+		bindMarketLinkEvents: function(parent, template) {
+			const self = this;
+			const linkForm = template.find('#cluster_link_form');
+
+			monster.ui.validate(linkForm);
+
+			parent.find('.market-link #save').on('click', function(e) {
+				e.preventDefault();
+
+				const saveThis = $(this);
+
+				saveThis.prop({ disabled: true });
+
+				if (monster.ui.valid(linkForm)) {
+					const formData = monster.ui.getFormData('cluster_link_form');
+
+					self.maybeSetApiUrl();
```

Comment:
If this potentially needs to run on every `app#updateMarketConnector`, how about moving it in that method body then?

Reply:
That makes a loop, `maybeSetApiUrl` also calls `updateMarketConnector` to set `api_url`.

Also the API is implement in a way that you can only perform a single action one at a time. You **_cannot_** set `enabled` and `api_url` in the same request.

If you really need those fancy update/post thing then you could always use `/system_config/appex_client` API directly. The `/appstore/marketplace` is a convenient way to configure appex_client without needing to activate `/system_config` crossbar API.

Also there is only few actions here, `link`/`unlink`, `enable` and `api_url`. We just want to make sure we set api_url when you're enabling and/or linking your cluster to make sure the appex client is functional without requiring you to set it manually when you enable and linking your cluster.

Reply:
Yep, that's why we usually wrap request handlers in their own methods and create yet other methods to wrap higher levels of abstraction (e.g. enableMarketConnect, linkMarktConnector, ...).